File: services/api/app/schemas/device.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.device import DeviceOnlineStatus


class DeviceChannelRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    channel_no: int
    name: str
    online_status: DeviceOnlineStatus
    is_encrypted: bool | None
    video_level: int | None
    is_present: bool
    last_online_at: datetime | None
    last_synced_at: datetime


class DeviceRead(BaseModel):
    id: int
    provider: str
    serial_masked: str
    name: str
    model: str | None
    online_status: DeviceOnlineStatus
    is_encrypted: bool | None
    channel_count: int
    is_present: bool
    last_online_at: datetime | None
    last_synced_at: datetime
    channels: list[DeviceChannelRead]


class DeviceStatusRead(BaseModel):
    id: int
    serial_masked: str
    online_status: DeviceOnlineStatus
    is_encrypted: bool | None
    is_present: bool
    last_online_at: datetime | None
    last_synced_at: datetime


class DeviceSyncResponse(BaseModel):
    created: int
    updated: int
    missing: int
    channel_count: int
    synced_at: datetime


class EzvizIntegrationStatus(BaseModel):
    configured: bool
    token_cache: str
    token_status: str
    token_expires_at: datetime | None
    token_refreshed_at: datetime | None
    device_count: int
    online_count: int
    last_synced_at: datetime | None

File: apps/web/src/types/device.ts
export type DeviceOnlineStatus = 'online' | 'offline' | 'unknown'

export interface DeviceChannel {
  id: number
  channel_no: number
  name: string
  online_status: DeviceOnlineStatus
  is_encrypted: boolean | null
  video_level: number | null
  is_present: boolean
  last_online_at: string | null
  last_synced_at: string
}

export interface Device {
  id: number
  provider: string
  serial_masked: string
  name: string
  model: string | null
  online_status: DeviceOnlineStatus
  is_encrypted: boolean | null
  channel_count: number
  is_present: boolean
  last_online_at: string | null
  last_synced_at: string
  channels: DeviceChannel[]
}

export interface DeviceSyncResult {
  created: number
  updated: number
  missing: number
  channel_count: number
  synced_at: string
}

export interface DeviceStatus {
  id: number
  serial_masked: string
  online_status: DeviceOnlineStatus
  is_encrypted: boolean | null
  is_present: boolean
  last_online_at: string | null
  last_synced_at: string
}

export interface EzvizIntegrationStatus {
  configured: boolean
  token_cache: 'memory' | 'redis'
  token_status: 'not_configured' | 'not_cached' | 'valid' | 'refresh_required'
  token_expires_at: string | null
  token_refreshed_at: string | null
  device_count: number
  online_count: number
  last_synced_at: string | null
}
